---
layout: post
title: Paper Summary Cussat-Blanc et al. 2015
---

### Paper Summary -  “Gene Regulatory Network Evolution Through Augmenting Topologies”

The goal of this paper is to present an efficient Evolutionary Algorithm (EA) for evolving high quality Gene Regulatory Networks (GRNs). Genetic Algorithms (GAs) are often used to evolve GRNs and the authors propose that a GA specifically adapted to GRNs will achieve these efficiency and quality goals. Improving the quality of the solution is done by improving “evolvability” or “the capacity to generate heritable phenotypic properties by reducing the potential lethality of variations, and/or reducing the number of reproductive events required to reach a novel trait.” Their algorithm, called GRNEAT (GRN Evolution by Augmenting Topology), is inspired by the NEAT (NeuroEvolution of Augmenting Topologies) algorithm, which is for evolving neural networks. 

The GRN model of Cussat-Blanc et al. is a set of proteins and a few formula coefficients, each with a type and a 3 tags. The type indicates whether it is an input protein, an output protein, or a regulatory protein. This dictates whether a protein can be affected by other proteins and vice versa, where a protein being affected means a change in its concentration. Specifically, an input protein regulates other proteins but is not regulated by them, an output protein is regulated by other proteins but does not regulate them, and a regulatory (or internal) protein both regulates and is regulated by other proteins. The 3 tags are a protein tag, an enhancer tag, and an inhibitor tag; all of them are integers in a finite chosen range. The protein tag can be thought of as the protein sequence: it is used to calculate matching to other protein’s enhancer and inhibitor tags when determining what effect they will have on this protein’s concentration. This concentration change calculation uses the coefficient included in the GRN  and considers all pairwise comparisons between the current protein’s protein tag and the enhancer and inhibitor tags on all others relevant proteins.

By creating a mapping function this model can interface with any problem domain. This entails mapping a problem’s input values onto protein concentrations and after evolution, mapping protein concentrations onto problem output values to get a solution. In this model protein concentrations are any rational number between 0 and 1 inclusive.

Cussat-Blanc et al. draw from an algorithm for neural networks because they have the same problem as GRNs in that a small change to a part of the system can have a dramatic and somewhat unpredictable impact on the whole, which is particularly a problem in the case of a crossover operation. The NEAT algorithm that inspired their algorithm and was able to solve this crossing over problem in neural nets has 3 key characteristics that the authors adapted to GRNs. First, instead of randomly initializing the population using the whole search space, only small individuals are generated so that the population will grow in complexity progressively. Second, they use a speciation mechanism that group similar genomes to afford some temporary protection to new rarer traits during evolution. Third, they use an alignment process in their crossover operator to avoid offspring with missing important genetic content.

To initialize their population with small GRNs they generate one input protein for every problem input, one output protein for every problem output, and one regulatory protein, each with randomly generated protein, enhancer, and inhibitor tags. The authors state that this is the minimum structure and genome size required for functional GRNs in their model. Then, to create initial species of a useful evolvable size, they duplicate each initial genome by applying a mutation to one of the proteins to get a group of twice the minimum species size. By creating all individuals in the species by mutation from one initial genome the generated GRNs are similar enough to be part of the same species, and by generating double the amount of the minimum species size they get enough variety in the species to have diverse offspring and therefore useful exploration of the search space.

Since a species is a grouping of similar genomes a measure of similarity or genetic distance is required to determine species memberships and speciation events, where a subgroup of individuals split into their own distinct species group. The distance between proteins is calculated using a weighted sum of the difference between matching tags divided by a model precision factor. The protein tag has significantly more weight than the enhancer and inhibitor tags. Using this formula for protein distance the distance between two GRNs is calculated by summing a few distances then normalizing by the number of proteins in the larger genome. These distances are: the distance between corresponding input and output proteins, the minimum pairwise distance from regulatory proteins in one GRN to regulatory proteins in the other GRN with replacement, and the distance between GRN coefficients. 

Using this distance function, species are then organized around a representative individual, which is either the first individual in a new species or a randomly selected species member from the previous generation. The authors elected to use a random selection for the representative since this is less computationally expensive than other selection methods and produces good results regardless. An individual is included in a species if the distance between them and the representative is less than that species’ speciation threshold, which is bigger for smaller species and smaller for bigger species and is adjusted dynamically between value of 0.03 and 0.5 inclusive. 

After each new generation is created, individuals are reorganized into species using this method. If a species ever falls beneath a minimum species size, then the species and all its members are deleted and the population is replenished by generating random genomes in other species. When doing this, the parent is chosen using tournament selection and the new genome is created by mutation. However, this new mutated genome has to fit into the parent’s species, so new genomes will be generated by mutation from this parent until one is found that is a member of the species.

Selection and reproduction happen within each species, but the species compete with each other for offspring since the number of offspring allocated to each species depends on their size and overall fitness of members. An individual’s fitness score is adjusted by dividing by the size of their species and the number of offspring a species gets is the sum of the adjusted fitnesses of its members over the total adjusted fitnesses of the population. Once this number has been determined, the next generation is generated by a combination of selection and reproduction in each species. First, the fittest individual from a species is automatically included in the next generation. Then, the mutation and crossover rate global parameters along with the allotted number of offspring determine how many children should be generated with each method. If the sum of these rates is inferior to 1, then additional individuals from the parent generation are selected using a standard 3-player tournament method.

GRNEAT uses 2 genetic operators for reproduction: mutation and crossovers. When a genome is mutated, either one tag of a random protein is modified, a new randomly generated regulatory protein is added, or one random regulatory protein is deleted. Which of these mutations is applied is randomly selected with 0.5 probability for adding a protein and 0.25 probability each for mutating and deleting proteins. The authors found that these values provided an ideal balance between exploration and slowly increasing GRN size.

The alignment-based crossover operator for GRNs is a significant contribution of Cussat-Blanc et al. This operator crosses each type of protein separately. Input and output proteins are treated the same: one protein corresponding to each problem input/output is randomly selected from one parent, so the 2 resulting children have complementary sets. As for the regulatory proteins, they are shuffled, aligned, and then one protein from each pair is randomly selected. Alignment is based on the previously mentioned protein distance and uses a threshold. Proteins from one parent are aligned to the closest match that passes the alignment threshold in the other parent. Each protein can only be aligned once. In the case that there are unaligned proteins, the child get all the unaligned proteins from one parent. Which parent’s proteins they get depends on how many proteins they have already inherited from that parent. This alignment and allocation of unaligned proteins helps ensure that no important genetic material goes missing from a child and that unaligned proteins are better suited to interacting with the aligned proteins inherited. The coefficients are also crossed by random selection from one parent.

The authors compared the performance of GRNEAT to a standard GA and Evolutionary Programming (EP) algorithms, GA with speciation mechanic, GA with aligned crossover, and EP with speciation mechanic, on a variety of problems. Their results were similar across all experiments. GRNEAT is slow to start, likely due to its initialization, but afterwards it generally produces solutions with better fitness than other the algorithms tested. GRNEAT is significantly better than the standard GA in all problems but similar in performance to the standard EP in a few of the problems. The GA with aligning crossover also often attained similar fitness performance but with more complex solutions in terms of GRN size than GRNEAT. As a general trend, speciation helped to decrease the size of solutions and aligning crossover contributed to developing solutions with better fitness. The authors also tested the developed GRNs on data not in their training set to evaluate the generalization ability of the solutions and found that GRNEAT solutions were significantly better than or similar to the next best result.

The authors conclude that their GRNEAT algorithm is more suited to evolving GRNs than other existing methods. It achieves this by improving their evolvability through a specialized initialization process, an aligning crossover operator, and a speciation process, all inspired by the NEAT algorithm. They suggest that further research could be done on any one of these 3 key features, but emphasize that research on how to manage species size and other clustering methods might be worth investigating.


[Cussat-Blanc et al., 2015] S. Cussat-Blanc, K. Harrington and J. Pollack, "Gene Regulatory Network Evolution Through Augmenting Topologies", IEEE Transactions on Evolutionary Computation, vol. 19, no. 6, pp. 823-837, 2015. Available: 10.1109/tevc.2015.2396199 
